04_basic_struct/main.c

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: MIT and GPL
+#include <linux/init.h>
+#include <linux/module.h>
+#include <linux/device.h>
+#include <linux/err.h>
+
+#include <linux/sysfs.h>
+#include <linux/device.h>
+
+#include <linux/types.h>
+#include <linux/list.h>
+#include <linux/slab.h>
+#include <linux/string.h>
+
+static struct kobject *hello;
+
+static LIST_HEAD(list);
+
+struct hello_node {
+	struct list_head head;
+	char *str;
+};
+
+static ssize_t rw_show(struct kobject *kobj,
+				struct kobj_attribute *attr, char *buf)
+{
+	struct hello_node *item = NULL;
+	int offset = 0;
+	int state = 0;
+
+	list_for_each_entry(item, &list, head) {
+		pr_info("Try to read some info\n");
+		state = sprintf(buf + offset, "%s", item->str);
+		if (state < 0) {
+			pr_info("Error\n");
+			return state;
+		}
+		pr_info("Success\n");
+		offset += state;
+	}
+	return offset;
+}
+
+static ssize_t rw_store(struct kobject *kobj,
+		struct kobj_attribute *attr, const char *buf, size_t count)
+{
+	struct hello_node *item = kzalloc(sizeof(*item), GFP_KERNEL);
+
+	if (strlen(buf) != count)
+		return -EINVAL;
+	item->str = kmalloc_array(count, sizeof(char), GFP_KERNEL);
+	strcpy(item->str, buf);
+	list_add(&item->head, &list);
+	pr_info("New item in list %s\n", buf);
+	return count;
+}
+
+static struct kobj_attribute attr =
+	__ATTR(list, 0660, rw_show, rw_store);
+
+static int mymodule_init(void)
+{
+	int state = 0;
+
+	pr_info("Module initialized successfully\n");
+	hello = kobject_create_and_add("kobject_list", kernel_kobj);
+	if (hello == NULL) {
+		pr_info("impossible to create Kobject!\n");
+		state = -EINVAL;
+	} else {
+		state = sysfs_create_file(hello, &attr.attr);
+		if (state == -EINVAL) {
+			kobject_put(hello);
+			pr_info("impossible to create File!\n");
+		}
+	}
+	return state;
```

Comment:
better to use return on errors when possible
this way you can simplify the code
remove "else" and brackets and just put following code on the same indentation level